apm-lambda-extension/extension/route_handlers.go

```diff
@@ -30,18 +32,28 @@ type AgentData struct {
 }
 
 var AgentDoneSignal chan struct{}
+var mainExtensionContext context.Context
 
 // URL: http://server/
-func handleInfoRequest(apmServerUrl string) func(w http.ResponseWriter, r *http.Request) {
+func handleInfoRequest(ctx context.Context, apmServerUrl string, config *extensionConfig) func(w http.ResponseWriter, r *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
 
+		Log.Debug("Handling APM Server Info Request")
+		mainExtensionContext = ctx
+
 		// Init reverse proxy
 		parsedApmServerUrl, err := url.Parse(apmServerUrl)
 		if err != nil {
 			Log.Errorf("could not parse APM server URL: %v", err)
 			return
 		}
+
 		reverseProxy := httputil.NewSingleHostReverseProxy(parsedApmServerUrl)
+		reverseProxyTimeout := time.Duration(config.DataForwarderTimeoutSeconds) * time.Second
+		reverseProxy.Transport = &http.Transport{
+			ResponseHeaderTimeout: reverseProxyTimeout,
+		}
```

Comment:
As mentioned offline, please do not overwrite the whole `http.Transport` here, but rather only set the `ResponseHeaderTimeout`, or even pass it as an option above when calling `httputil.NewSingleHostReverseProxy`.

Reply:
After both reviewing the [implementation](https://cs.opensource.google/go/go/+/master:src/net/http/httputil/reverseproxy.go;drc=a10a209b23f877c80d8a5f3ebda1ce4b492ac3a9;l=144) of `NewSingleHostReverseProxy` and debugging the current implementation, it seems that the the default reverse proxy comes with a `nil` transport, which then gets [processed](https://cs.opensource.google/go/go/+/master:src/net/http/httputil/reverseproxy.go;drc=a10a209b23f877c80d8a5f3ebda1ce4b492ac3a9;l=215) as `http.DefaultTransport` by `ServeHTTP()`.

![image](https://user-images.githubusercontent.com/48380853/165115903-35ca3b24-d046-4501-a519-72f460519b76.png)

To alleviate the fears discussed online, I have [replaced](
2240711) the creation of a `Transport` from scratch by a modification of `http.DefaultTransport`, which is the Transport that is used anyway.